[CO-389] Write properties for cardano-sl-x509

KtorZ opened the pull request:
## Description

As per comment in #3610 --> https://github.com/input-output-hk/cardano-sl/pull/3610#issuecomment-421955739

## Linked issue

[[CO-379]](https://iohk.myjetbrains.com/youtrack/issue/CO-389)

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- ~~[ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.~~

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

KtorZ commented on x509/test/Test/Cardano/X509/Configuration/Arbitrary.hs:
```diff
@@ -0,0 +1,84 @@
+        <*> arbitraryBasicString
+        <*> oneof [pure Nothing, Just <$> arbitraryBasicString]
+
+    shrink _ = []
```
Well, not really right? There are shrinkers defined when they make sense. Here, there's just no way to shrink a `DirConfiguration`. Nevertheless, even without shrinking, a `DirConfiguration` don't get really unmanageable :sweat_smile:

KtorZ replied:
That's actually why I have also _explictely defined_ it as `[]`, to make it clear that I haven't forgotten to define a `shrink` method for that type but that is just that this is the only sensible choice :) 

And so far, failures render nicely :stuck_out_tongue: (I got a few during development)

disassembler commented on x509/src/Cardano/X509/Configuration.hs:
```diff
@@ -304,27 +330,22 @@ usExtensionsV3 purpose subDN issDN =
+parseAltName :: MonadThrow m => String -> m AltName
+parseAltName name =
+    case IP.decode (T.pack name) of
+        Just _  -> throwM (ErrInvalidAltName "IP was given as an AltName but it should be a DNS name")
```
This will break deadalus. Please leave the IP SAN parsing in.

KtorZ replied:
Hmmm... Good catch @disasm 
But IP SAN aren't _recognized as a valid alternative names_ by the x509 library we use. I can perhaps tweak the hook which validates the name...

Does Daedalus uses 127.0.0.1 to contact the server instead of `localhost`?

disassembler commented on x509/test/Test/Cardano/X509/Configuration/Arbitrary.hs:
```diff
@@ -0,0 +1,155 @@
+instance Arbitrary (Invalid AltNames) where
+    arbitrary =
+        fmap (Invalid . mkAltNames) $ listOf1 $ frequency
+            [ (80, pure "localhost")
```
Does this mean localhost is an invalid altName as well now? That breaks wallet connectScripts if so.

KtorZ replied:
No. The `Invalid` is just biaised to sometimes generate invalid configurations such that we can use it to trigger multiple sort of errors.

KtorZ commented:
@disassembler 

Okay, I've updated the implementation to:

- Accept IPs as SAN (using the IPv4 and IPv6 encoding / decoding functions)
- Properly validate IP SAN and DNS SAN. The former one just does a plain equality check and the latter relies on the default validation hook from x509-validation (which validates DNS SAN or fails)
- Updated properties to reflect these changes
- Have a new property which checks that unknown ServiceID (i.e. not defined as SAN) are actually rejected, for server certificates (we don't check the name in case of client certificates because it doesn't make sense).

Let me know :+1:

Also: I am not sure whether I should add something to `release.nix` or anything. Note that this package is going to be used by `cardano-sl-wallet-new` (PR incoming) so it will get built anyway. Would tests run as part of the build automatically?

KtorZ commented:
@disassembler 
So, while using this on the "demo cluster" story, I noticed an issue with our TLS Manager with IP SANs. We basically need to have our custom name validator here as well, so I've added it:

https://github.com/input-output-hk/cardano-sl/pull/3620/files#diff-a879fd32cda407f37cf36a9459d628e8
